src/lib/SoftHSM.cpp

```diff
@@ -11383,23 +11524,76 @@ CK_RV SoftHSM::deriveSymmetric
+				// attributes set to CK_TRUE.
+				bool bAlwaysSensitive = baseKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false) &&
+										otherKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false);
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bAlwaysSensitive);
+				// The derived key’s CKA_NEVER_EXTRACTABLE attribute is set to CK_TRUE
+				// if and only if both of the original keys have their CKA_NEVER_EXTRACTABLE
+				// attributes set to CK_TRUE
+				bool bNeverExtractable = baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false) &&
+										 otherKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false);
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bNeverExtractable);
+			}
+			else if (pMechanism->mechanism == CKM_CONCATENATE_BASE_AND_DATA ||
+				 pMechanism->mechanism == CKM_CONCATENATE_DATA_AND_BASE)
+			{
+				// [PKCS#11 v2.40, 2.31.4-2.31.7]
+				// If the base key has its CKA_SENSITIVE attribute set to CK_TRUE, so does the derived key.
+				// If not, then the derived key’s CKA_SENSITIVE attribute is set either from the supplied
+				// template or from a default value.
+				if (baseKey->getBooleanValue(CKA_SENSITIVE, true)) {
+					bOK = bOK && osobject->setAttribute(CKA_SENSITIVE, true);
+				}
+				// If the base key has its CKA_EXTRACTABLE attribute set to CK_FALSE, so does the derived key.
+				// If not, then the derived key’s CKA_EXTRACTABLE attribute is set either from the supplied
+				// template or from a default value.
+				if (!baseKey->getBooleanValue(CKA_EXTRACTABLE, false)) {
+					bOK = bOK && osobject->setAttribute(CKA_EXTRACTABLE, false);
+				}
+				// The derived key’s CKA_ALWAYS_SENSITIVE attribute is set to CK_TRUE if and only
+				// if the base key has its CKA_ALWAYS_SENSITIVE attribute set to CK_TRUE.
+				bool bAlwaysSensitive = baseKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false);
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bAlwaysSensitive);
+				// The derived key’s CKA_NEVER_EXTRACTABLE attribute is set to CK_TRUE if and only
+				// if the base key has its CKA_NEVER_EXTRACTABLE attribute set to CK_TRUE.
+				bool bNeverExtractable = baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false);
+				bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE, bNeverExtractable);
+			} else {
+				if (baseKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false)) {
+					bool bAlwaysSensitive = osobject->getBooleanValue(CKA_SENSITIVE, false);
+					bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bAlwaysSensitive);
+				} else {
+					bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, false);
+				}
+				if (baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, true)) {
+					bool bNeverExtractable = osobject->getBooleanValue(CKA_EXTRACTABLE, false) == false;
+					bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE, bNeverExtractable);
+				} else {
+					bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE, false);
+				}
```

Comment:
What cases does this else branch cover?

Reply:
It's a default branch for other mechanisms.

Reply:
Sorry, did not notice that this code is just moved into else branch. I thought it was added, which seemed suspicious. OK then.